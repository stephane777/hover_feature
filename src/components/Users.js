import React from "react";
import UserList from "./UserList";
import UserCard from "./UserCard";
import PropTypes from "prop-types";

const userList = (user) => {
	const props = {
		key: user.login.uuid,
		data: user,
	};
	return <UserList {...props} />;
};

const userCard = (user) => {
	const props = {
		key: user.login.uuid,
		data: user,
	};
	return <UserCard {...props} />;
};

const Users = ({ users, route }) => {
	return users.map((user) => {
		return route === "userList" ? userList(user) : userCard(user);
	});
};

Users.propTypes = {
	users: PropTypes.array.isRequired,
	route: PropTypes.oneOf(["userList", "userCard"]),
};

Users.defaultProps = {
	route: "userList",
};
export default Users;
