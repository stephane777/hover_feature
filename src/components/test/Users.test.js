import React from "react";
import Users from "../Users";
import { render, screen, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import renderer, { act } from "react-test-renderer";

import { twoUsers } from "./utils";

describe("test User component", () => {
	it("should render the User component with default userList props", () => {
		let mainContainer;
		expect.assertions(1);
		act(() => {
			const { container } = render(<Users users={twoUsers} />);
			mainContainer = container;
		});
		expect(mainContainer.firstChild.firstChild.className).toBe("userList");
		// expect()
	});
	it("should show an image, title, first and lastname", () => {
		expect.assertions(23);
		act(() => {
			render(<Users users={twoUsers} />);
		});
		expect(screen.getAllByRole("img").length).toBe(2);
		twoUsers.forEach(
			({
				login: { uuid },
				name: { title, first, last },
				picture: { thumbnail },
				dob: { age },
				email,
				phone,
				location: {
					street: { number, name },
					postcode,
					city,
				},
			}) => {
				const userContainer = document.getElementById(uuid);
				const utils = within(userContainer);
				const img = utils.getByRole("img");
				expect(img).toHaveAttribute("src", thumbnail);
				expect(img).toHaveAttribute("alt", `${title} ${first} ${last}`);
				expect(utils.getByText(title)).toBeInTheDocument();
				expect(utils.getByText(first)).toBeInTheDocument();
				expect(utils.getByText(last)).toBeInTheDocument();

				expect(utils.queryByText(age)).toBeNull();
				expect(utils.queryByText(email)).toBeNull();
				expect(utils.queryByText(phone)).toBeNull();
				expect(utils.queryByText(`${number} ${name}`)).toBeNull();
				expect(utils.queryByText(postcode)).toBeNull();
				expect(utils.queryByText(city)).toBeNull();
			}
		);
	});

	it("Should render user card: ", () => {
		const { container } = render(<Users users={twoUsers} route="userCard" />);
		expect(container.firstChild.className).toBe("userCard");
	});
	it("should render all user details", () => {
		expect.assertions(22);
		act(() => {
			render(<Users users={twoUsers} route="userCard" />);
		});
		twoUsers.forEach(
			({
				login: { uuid },
				name: { title, first, last },
				picture: { thumbnail },
				dob: { age },
				email,
				phone,
				location: {
					street: { number, name },
					postcode,
					city,
				},
			}) => {
				const userContainer = document.getElementById(uuid);
				const utils = within(userContainer);
				const img = utils.getByRole("img");
				expect(img).toHaveAttribute("src", thumbnail);
				expect(img).toHaveAttribute("alt", `${title} ${first} ${last}`);
				expect(utils.getByText(title)).toBeInTheDocument();
				expect(utils.getByText(first)).toBeInTheDocument();
				expect(utils.getByText(last)).toBeInTheDocument();

				expect(utils.getByText(age)).toBeInTheDocument();
				expect(utils.getByText(email)).toBeInTheDocument();
				expect(utils.getByText(phone)).toBeInTheDocument();
				expect(utils.getByText(`${number} ${name}`)).toBeInTheDocument();
				expect(utils.getByText(postcode)).toBeInTheDocument();
				expect(utils.getByText(city)).toBeInTheDocument();
			}
		);
	});

	describe("Snapshot testing", () => {
		let root;
		it("should match snapshot with userList", async () => {
			act(() => {
				root = renderer.create(<Users users={twoUsers} />);
			});
			expect(await root.toJSON()).toMatchSnapshot();
		});
		it("should match snapshot with userCard", async () => {
			act(() => {
				root.update(<Users users={twoUsers} route="userCard" />);
			});
			expect(await root.toJSON()).toMatchSnapshot();
		});
	});
});
