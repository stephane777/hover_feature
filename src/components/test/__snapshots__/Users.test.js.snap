// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test User component Snapshot testing should match snapshot with userCard 1`] = `
Array [
  <div
    className="userCard"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      id="488edda9-f042-434f-8162-e4dddf7055f5"
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "column",
          "fontSize": "1.4rem",
          "height": "100%",
          "justifyContent": "space-around",
          "overflow": "hidden",
          "width": "100%",
        }
      }
    >
      <div
        className="userCard__header u-margin-bottom-medium"
      >
        <img
          alt="Mrs Emily Mortensen"
          src="https://randomuser.me/api/portraits/thumb/women/78.jpg"
        />
        <span
          className="user__title"
        >
          Mrs
        </span>
        <span
          className="user__first"
        >
          Emily
        </span>
        <span
          className="user__last"
        >
          Mortensen
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          Age: 
        </span>
        <span>
          31
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          email: 
        </span>
        <span>
          emily.mortensen@example.com
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          Phone: 
        </span>
        <span>
          94033936
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          Street: 
        </span>
        <span>
          8651 Tingstedet
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          Postcode: 
        </span>
        <span>
          98755
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          City: 
        </span>
        <span>
          Randers Nv
        </span>
      </div>
    </div>
  </div>,
  <div
    className="userCard"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      id="bdc71f4e-e874-4575-96b5-5cfc6959458e"
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "column",
          "fontSize": "1.4rem",
          "height": "100%",
          "justifyContent": "space-around",
          "overflow": "hidden",
          "width": "100%",
        }
      }
    >
      <div
        className="userCard__header u-margin-bottom-medium"
      >
        <img
          alt="Miss Nurdan Tazegül"
          src="https://randomuser.me/api/portraits/thumb/women/9.jpg"
        />
        <span
          className="user__title"
        >
          Miss
        </span>
        <span
          className="user__first"
        >
          Nurdan
        </span>
        <span
          className="user__last"
        >
          Tazegül
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          Age: 
        </span>
        <span>
          31
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          email: 
        </span>
        <span>
          nurdan.tazegul@example.com
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          Phone: 
        </span>
        <span>
          (012)-335-1433
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          Street: 
        </span>
        <span>
          6949 Şehitler Cd
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          Postcode: 
        </span>
        <span>
          91771
        </span>
      </div>
      <div
        className="u-margin-bottom-medium"
      >
        <span
          className="bold"
        >
          City: 
        </span>
        <span>
          Balıkesir
        </span>
      </div>
    </div>
  </div>,
]
`;

exports[`test User component Snapshot testing should match snapshot with userList 1`] = `
Array [
  <div
    className=""
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="userList"
      id="488edda9-f042-434f-8162-e4dddf7055f5"
    >
      <img
        alt="Mrs Emily Mortensen"
        src="https://randomuser.me/api/portraits/thumb/women/78.jpg"
      />
      <span
        className="user__title"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Mrs
      </span>
      <span
        className="user__first"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Emily
      </span>
      <span
        className="user__last"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Mortensen
      </span>
    </div>
  </div>,
  <div
    className=""
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="userList"
      id="bdc71f4e-e874-4575-96b5-5cfc6959458e"
    >
      <img
        alt="Miss Nurdan Tazegül"
        src="https://randomuser.me/api/portraits/thumb/women/9.jpg"
      />
      <span
        className="user__title"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Miss
      </span>
      <span
        className="user__first"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Nurdan
      </span>
      <span
        className="user__last"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Tazegül
      </span>
    </div>
  </div>,
]
`;
